src/coreclr/vm/jitinterface.cpp

```diff
@@ -4634,6 +4634,30 @@ bool CEEInfo::isExactType(CORINFO_CLASS_HANDLE cls)
     return result;
 }
 
+// Returns whether a class handle represents a generic type, if that can be statically determined.
+TypeCompareState CEEInfo::isGenericType(CORINFO_CLASS_HANDLE cls)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    TypeHandle typeHandle = TypeHandle();
+
+    TypeCompareState result = TypeCompareState::May;
+
+    JIT_TO_EE_TRANSITION();    
+
+    if (!typeHandle.IsCanonicalSubtype())
```

Comment:
```suggestion
    if (typeHandle != TypeHandle(g_pCanonMethodTableClass))
```
IsCanonicalSubtype is too conservative condition

Reply:
I can revert that, sure! Just so I understand, what's going to guarantee that things like `typeof(T[]).IsGenericType` etc. will work correctly here with shared generics? I think that's the case @EgorBo mentioned on Discord suggesting to replace that `__Canon` check with that helper method instead. I'm probably missing something 😅

Reply:
> typeof(T[]).IsGenericType

`typeHandle` is going `__Canon[]` and this method is going to return false.

Reply:
The generic parameters are replaced substituded with `__Canon` when compiling shared generic code, but the type "structure" is otherwise preserved.